darshan-util/pydarshan/darshan/tests/test_glob_feature.py

```diff
@@ -0,0 +1,305 @@
+# Note: Some tests may currently fail, as this script is still under active development.
+# The log files here are from the the darshan-logs repository
+
+import sys
+import os
+import darshan
+from darshan.log_utils import get_log_path
+import pandas as pd
+from pandas.testing import assert_frame_equal
+import pytest
+import re
+print(sys.path)
+from darshan.glob_feature import glob_feature
+
+@pytest.mark.parametrize("log_name, expected_df", [
+     # grow this with more logs...
+     ("e3sm_io_heatmap_only.darshan",
+      pd.DataFrame({"filename_glob":
+                   ["/projects/radix-io/snyder/e3sm/can_I_out_h[.*].nc",
+                    "/projects/radix-io/E3SM-IO-inputs/i_case_1344p.nc"],
+                    "glob_count": [2, 1]})),
+
+      ("snyder_acme.exe_id1253318_9-27-24239-1515303144625770178_2.darshan",
+      pd.DataFrame({"filename_glob":
+                   ["/gpfs/mira-fs1/projects/radix-io/snyder/acme/test_F_case_cetus_dxt/run/[.*]",
+                    "/projects/ccsm/inputdata/atm/cam/chem/trop_mozart_aero/emis/[.*].nc",
+                    "/projects/ccsm/inputdata/atm/cam/physprops/[.*].nc",
+                    "/projects/ccsm/inputdata/atm/cam/[.*].nc",
+                    "/gpfs/mira-fs1/projects/radix-io/snyder/acme/test_F_case_cetus_dxt/run/test_F_case_cetus_dxt.[.*]00[.*]",
+                    "/projects/ccsm/inputdata/lnd/clm2/[.*].nc",
+                    "/gpfs/mira-fs1/projects/radix-io/snyder/acme/test_F_case_cetus_dxt/run/timing/[.*]i[.*]",
+                    "/projects/radix-io/snyder/acme/test_F_case_cetus_dxt/run/[.*].log.170927-064246",
+                    "/projects/ccsm/inputdata/atm/waccm/[.*].nc",
+                    "/projects/ccsm/inputdata/[.*]n.[.*].[.*]1[.*]0[.*].nc"], #Note: for this set of grouped paths it might be more benifical to display the individual filepaths 
+                    "glob_count": [22, 18, 14, 13, 10, 6, 6, 5, 3, 3]})),
+
+     ("darshan-apmpi-2nodes-64mpi.darshan",
+      pd.DataFrame({"filename_glob":
+                   ["/lus/theta-fs0/projects/Performance/chunduri/MILC/milctestv2-papi-reorder-darshan/MILC_2_526820_2021-06-14-15:58:47/[.*]n[.*]"],
+                    "glob_count": [2]})),
+
+     ("mpi-io-test.darshan",
+      pd.DataFrame({"filename_glob":
+                   ["/global/cscratch1/sd/ssnyder/tmp/mpi-io-test.tmp.dat"],
+                    "glob_count": [1]})),
+
+     ("e3sm_io_heatmap_and_dxt.darshan",
+      pd.DataFrame({"filename_glob":
+                   ["/projects/radix-io/snyder/e3sm/can_I_out_h[.*].nc",
+                    "/projects/radix-io/E3SM-IO-inputs/i_case_1344p.nc"],
+                    "glob_count": [2, 1]})),
+
+
+     ("hdf5_diagonal_write_1_byte_dxt.darshan",
```

Comment:
Taking this log file as an example of something you might want to get feedback from the Argonne team on.

The original DataFrame seems to be this:

<details>

```
6329959920961583796                                  /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/__init__.py
12081676647536809530         /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/__pycache__/__init__.cpython-38.pyc
9019270385924777942       /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_errors.cpython-38-x86_64-linux-gnu.so
17233738746347876168                                  /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/version.py
10563276375451711199          /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/__pycache__/version.cpython-38.pyc
16409530009729583527           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5.cpython-38-x86_64-linux-gnu.so
13248362266925776299         /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/defs.cpython-38-x86_64-linux-gnu.so
4638787767486382508      /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_objects.cpython-38-x86_64-linux-gnu.so
10170641380995794425                            /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5py_warnings.py
8697898997669214892     /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/__pycache__/h5py_warnings.cpython-38.pyc
7051742911845937566         /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_conv.cpython-38-x86_64-linux-gnu.so
9236461081719577833           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5r.cpython-38-x86_64-linux-gnu.so
2181399625010544346           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5t.cpython-38-x86_64-linux-gnu.so
6528078802042034082           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5p.cpython-38-x86_64-linux-gnu.so
2575676175211135484           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5s.cpython-38-x86_64-linux-gnu.so
11727641573666397676        /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/utils.cpython-38-x86_64-linux-gnu.so
17901672647343149144         /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5ac.cpython-38-x86_64-linux-gnu.so
9518436409801048664                                             /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/logging/__init__.py
2863652681302921111                     /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/logging/__pycache__/__init__.cpython-38.pyc
2204154745069321427                                                    /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/traceback.py
555215874445445411                             /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/__pycache__/traceback.cpython-38.pyc
3096762644015249429                                                       /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/string.py
3433981731425580030                               /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/__pycache__/string.cpython-38.pyc
6622959456894657244           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5z.cpython-38-x86_64-linux-gnu.so
16198910964451933166          /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5a.cpython-38-x86_64-linux-gnu.so
12844965888207067364       /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_proxy.cpython-38-x86_64-linux-gnu.so
8320392657652582313           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5d.cpython-38-x86_64-linux-gnu.so
2182783899281604896          /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5ds.cpython-38-x86_64-linux-gnu.so
6824284610368689248           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5f.cpython-38-x86_64-linux-gnu.so
2255611827735696210           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5g.cpython-38-x86_64-linux-gnu.so
7369759469852332435           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5i.cpython-38-x86_64-linux-gnu.so
3786189409375561029          /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5fd.cpython-38-x86_64-linux-gnu.so
9453911172532451432          /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5pl.cpython-38-x86_64-linux-gnu.so
7281569168338684097                              /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__init__.py
6456765083458423379      /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/__init__.cpython-38.pyc
1619390675886412049                               /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/filters.py
6004586721157325469       /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/filters.cpython-38.pyc
7590428293607185835                                /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/compat.py
941427344463918395         /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/compat.cpython-38.pyc
7880173344308596059                                  /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/base.py
412714883876212763           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/base.cpython-38.pyc
8629723084015159315                                 /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/files.py
1259607287293063192         /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/files.cpython-38.pyc
1754798185250444829                                 /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/group.py
14745068672433573323        /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/group.cpython-38.pyc
8305533062347777536           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5o.cpython-38-x86_64-linux-gnu.so
15958169687914591130          /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5l.cpython-38-x86_64-linux-gnu.so
11692197895731416386                              /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/dataset.py
18374151976672635155      /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/dataset.cpython-38.pyc
5947139205337478310     /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_selector.cpython-38-x86_64-linux-gnu.so
4123734041483639167                            /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/selections.py
17375372070410440550   /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/selections.cpython-38.pyc
11306757759007388579                          /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/selections2.py
5194722322225707588   /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/selections2.cpython-38.pyc
13359638461944338488                             /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/datatype.py
17244020578242913766     /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/datatype.cpython-38.pyc
16568253267388065553                                  /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/vds.py
8979639072945830616           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/vds.cpython-38.pyc
15318755126620508599                                                        /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/copy.py
12711741908924247827                                /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/__pycache__/copy.cpython-38.pyc
7714427140906174735                                 /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/attrs.py
9344690504834470365         /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/_hl/__pycache__/attrs.cpython-38.pyc
16918611628337716751                                                        /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/uuid.py
8511949921801793794                                 /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/__pycache__/uuid.cpython-38.pyc
2130744796581944128                                               /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_0.h5
17389075414465034664                                     /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_0.h5:/dataset
4223954931879612592                                               /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_1.h5
14104557419321271749                                     /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_1.h5:/dataset
3586708446079182075                                               /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_2.h5
14252080315496364280                                     /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_2.h5:/dataset
17916052597340637289                                              /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_3.h5
15211073457117916342                                     /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_3.h5:/dataset
9342247733453943503                                               /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_4.h5
10985895144053482380                                     /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_4.h5:/dataset
4898891139486729695                                               /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_5.h5
15357467061712271470                                     /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_5.h5:/dataset
3386331313084499192                                               /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_6.h5
409581163713740906                                       /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_6.h5:/dataset
7972149951920014792                                               /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_7.h5
2753966606541326699                                      /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_7.h5:/dataset
5562527676113658148                                               /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_8.h5
6263533231926069969                                      /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_8.h5:/dataset
8261993854308713163                                               /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_9.h5
1478452523172783126                                      /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_9.h5:/dataset
```

</details>

That table has file paths including the following:

```
8320392657652582313           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5d.cpython-38-x86_64-linux-gnu.so
2182783899281604896          /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5ds.cpython-38-x86_64-linux-gnu.so
6824284610368689248           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5f.cpython-38-x86_64-linux-gnu.so
2255611827735696210           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5g.cpython-38-x86_64-linux-gnu.so
7369759469852332435           /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5i.cpython-38-x86_64-linux-gnu.so
3786189409375561029          /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5fd.cpython-38-x86_64-linux-gnu.so
9453911172532451432          /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/h5pl.cpython-38-x86_64-linux-gnu.so
```

But the granularity in the test below is obviously coarser than clustering that group individually, so may be good to ask the Argonne team what they want in this case. Your actual and expected values do still differ here as well, so I'm curious if you are still maintaining that the "expected" value is the right one here?

```
actual_table:
                                                                            filename_glob  glob_count
0                      /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/[.*]          64
1  /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_[.*].h5[.*]          20
expected_df:
                                                                            filename_glob  glob_count
0   /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/site-packages/h5py/[.*]          54
1  /yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_[.*].h5[.*]          20
2                      /users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/[.*]          10
```

Reply:
Hopefully, it doesn't throw a wrench in anything, but after looking closer at the filenames for this example, I think I'd actually suggest something like this for expected output (ignoring the glob counts, I didn't implement this so don't have the final counts handy):

```
/users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/[.*].py
/users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/[.*].pyc
/users/nawtrey/.conda/envs/pydarshan_hdf5_py38/lib/python3.8/[.*].so
/yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_[.*].h5
```

So, at a high-level it tries to find the "common" directories that have lots of files in them, then within those common directories bins things based on suffix? I think that works well for this particular scenario, but maybe others have different thoughts. Of course, you'll have to have some algorithmic way to decide when to stop splitting into common subdirectories (and start trying to group into suffixes), if that makes sense.

Note, that I'm also ignoring the weird HDF5 record names that include the dataset (e.g., '/yellow/users/nawtrey/projects/hdf5_testing/test_files_write_1_bytes/test_9.h5:/dataset'). I still think we should explicitly ignore those. The only way I know how to do that is to restrict record names to only those coming from POSIX and STDIO modules:
- Read in POSIX and STDIO module data
- Remove all name records that don't have a corresponding match in the POSIX/STDIO record list
- Perform analysis on those records